pkg/archive/archive.go

```diff
@@ -449,9 +449,6 @@ func TarWithOptions(srcPath string, options *TarOptions) (io.ReadCloser, error)
 				}
 
 				if skip {
-					if f.IsDir() {
```

Comment:
Been thinking about the performance impact of this PR and I'm a little bit worried about the removal of this SkipDir stuff.  For very large build contexts this could have a serious impact on our upload speed - especially when you consider the fact that we do certain cleanup type of actions over and over in our Matches() func.  For example, we call filepath.Clean() on each pattern and we do it for all patterns and repeat that for each file in the build context.

What I'd like to propose is this....
* before this for-loop (before line 422) loop over options.ExcludePatterns. For each, call filepath.Clean() and save the results in a new slice - so we don't mess with the original.
* also save the split(...,"/") version of the pattern so we don't need to do it in Match()
* while looping, also detect any ! usage and make a note of it - actually, note if we're *not* using it at all.
* error any !'s all by themselves during this loop instead of in Match()
* if there are no !'s in any patterns then keep/do the `return filepath.SkipDir` on line 453.
* move Match() from fileutils (assuming its not used by anyone else) into archive - mainly because the optimizations I'm talking about here make this func less generic and will make some assumptions that a 'generic' pkg shouldn't make.  If Match() shrinks enough, perhaps even just move it into this func at line 444 instead - but worst case make it a sibling to this one.

thoughts?